gpii/node_modules/journal/src/Journal.js

```diff
@@ -0,0 +1,278 @@
+    listeners: {
+        "{lifecycleManager}.events.onSessionStart": {
+            namespace: "logRestoreUrl",
+            func: "gpii.journalApp.logRestoreUrl",
+            args: ["{that}", "{arguments}.0"]
+        }
+    },
+    templates: {
+        listJournals: "<html><head><title>Journals List</title></head><body><p>Click on a journal link to restore the system to its state at that time </p>%journals</body></html>",
+        journalLink: "<p><a href=\"%journalUrl\">System snapshot taken at %journalTime</a>%crashedString</p>"
+    },
+    requestHandlers: {
+        restore: {
+            route: "/journal/restore/:journalId",
+            method: "get",
+            type: "gpii.journal.restoreJournal.handler",
+        },
+        journals: {
+            route: "/journal/journals.html",
+            method: "get",
+            type: "gpii.journal.journals.handler"
+        }
+    }
+});
+
+gpii.journalApp.dateToRestoreUrl = function (that, server, date) {
+    var baseUrl = that.options.serverUrl + ":" + server.options.port;
+    var pathTemplate = that.options.requestHandlers.restore.route.replace(":", "%");
+    var journalId = ">" + new Date(date).toISOString();
+    var pathReplaced = kettle.dataSource.URL.resolveUrl(pathTemplate, {journalId: journalId});
+    var fullUrl = baseUrl + pathReplaced;
+    return fullUrl;
+};
+
+gpii.journalApp.logRestoreUrl = function (that, gradeName) {
+    if (gradeName === "gpii.lifecycleManager.userSession") {
+        var fullUrl = that.dateToRestoreUrl(Date.now());
+        fluid.log(fluid.logLevel.WARN, "New user session starting: In future, you can visit the URL " +
+            fullUrl + " to restore the system's settings as they were at this point");
+    }
+};
+
+fluid.defaults("gpii.journal.restoreJournal.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.journal.restoreJournal.handleRequest",
+            args: [ "{gpii.journal}", "{that}"]
+        }
+    }
+});
+
+gpii.journal.restoreJournal.handleRequest = function (journal, request) {
+    var journalId = request.req.params.journalId;
+    var restorePromise = journal.restoreJournal(journalId);
+    fluid.promise.follow(restorePromise, request.handlerPromise);
+};
+
+fluid.defaults("gpii.journal.journals.handler", {
+    gradeNames: ["kettle.request.http"],
+    invokers: {
+        handleRequest: {
+            funcName: "gpii.journal.journals.handleRequest",
+            args: [ "{gpii.journal}", "{that}"]
+        }
+    }
+});
+
+gpii.journal.journals.handleRequest = function (journal, request) {
+    var links = fluid.transform(journal.journalFiles, function (journalFile) {
+        var terms = {
+            journalUrl: journal.dateToRestoreUrl(journalFile.createTime),
+            journalTime: new Date(journalFile.createTime).toLocaleString(),
+            crashedString: journal.closed ? "" : ": crashed session"
```

Comment:
I think `journal.closed` should be `journalFile.closed` here. If I log in and log out again, I see that a journal file has been created with `"closed": true` as expected. But the web page list of journals shows "crashed session".